File: src/plugins/streaks.ts
import { IMatrixClient, Plugin, CommandDef, MessageContext } from "./base";
import { getDb } from "../db";

export class StreaksPlugin extends Plugin {
  constructor(client: IMatrixClient) {
    super(client);
  }

  get name() {
    return "streaks";
  }

  get commands(): CommandDef[] {
    return [
      { name: "streak", description: "Current and record streak", usage: "!streak [@user]" },
      { name: "firstboard", description: "Early bird leaderboard", usage: "!firstboard [@user|month]" },
    ];
  }

  async onMessage(ctx: MessageContext): Promise<void> {
    // Passive: track daily activity and first poster
    this.trackActivity(ctx);

    if (this.isCommand(ctx.body, "firstboard")) {
      await this.handleFirstboard(ctx);
    } else if (this.isCommand(ctx.body, "streak")) {
      await this.handleStreak(ctx);
    }
  }

  private trackActivity(ctx: MessageContext): void {
    const db = getDb();
    const today = new Date().toISOString().slice(0, 10);

    // Track daily activity
    db.prepare(`
      INSERT INTO daily_activity (user_id, room_id, date, message_count)
      VALUES (?, ?, ?, 1)
      ON CONFLICT(user_id, room_id, date) DO UPDATE SET message_count = message_count + 1
    `).run(ctx.sender, ctx.roomId, today);

    // Track first poster (INSERT OR IGNORE — first one wins)
    db.prepare(`
      INSERT OR IGNORE INTO daily_first (room_id, date, user_id, timestamp)
      VALUES (?, ?, ?, datetime('now'))
    `).run(ctx.roomId, today, ctx.sender);

    // Update streak in user_stats
    this.updateStreak(ctx.sender, ctx.roomId, today);
  }

  private updateStreak(userId: string, roomId: string, today: string): void {
    const db = getDb();
    const stats = db
      .prepare(`SELECT last_active_date, current_streak, longest_streak FROM user_stats WHERE user_id = ? AND room_id = ?`)
      .get(userId, roomId) as
      | { last_active_date: string | null; current_streak: number; longest_streak: number }
      | undefined;

    if (!stats) return;

    const lastDate = stats.last_active_date;
    if (lastDate === today) return; // Already counted today

    // Check if yesterday was active
    const yesterday = new Date();
    yesterday.setUTCDate(yesterday.getUTCDate() - 1);
    const yesterdayStr = yesterday.toISOString().slice(0, 10);

    let newStreak: number;
    if (lastDate === yesterdayStr) {
      newStreak = stats.current_streak + 1;
    } else {
      newStreak = 1;
    }

    const newLongest = Math.max(stats.longest_streak, newStreak);

    db.prepare(`
      UPDATE user_stats SET current_streak = ?, longest_streak = ?
      WHERE user_id = ? AND room_id = ?
    `).run(newStreak, newLongest, userId, roomId);
  }

  private async handleStreak(ctx: MessageContext): Promise<void> {
    const args = this.getArgs(ctx.body, "streak");
    const targetUser = args.startsWith("@") ? args.split(/\s/)[0] : ctx.sender;

    const db = getDb();
    const stats = db
      .prepare(`SELECT current_streak, longest_streak FROM user_stats WHERE user_id = ? AND room_id = ?`)
      .get(targetUser, ctx.roomId) as { current_streak: number; longest_streak: number } | undefined;

    if (!stats) {
      await this.sendReply(ctx.roomId, ctx.eventId, `No streak data for ${targetUser}.`);
      return;
    }

    await this.sendMessage(
      ctx.roomId,
      `${targetUser}'s streak: ${stats.current_streak} day${stats.current_streak !== 1 ? "s" : ""} (record: ${stats.longest_streak})`
    );
  }

  private async handleFirstboard(ctx: MessageContext): Promise<void> {
    const db = getDb();

    const rows = db
      .prepare(`
        SELECT user_id, COUNT(*) as firsts
        FROM daily_first
        WHERE room_id = ?
        GROUP BY user_id
        ORDER BY firsts DESC
        LIMIT 10
      `)
      .all(ctx.roomId) as { user_id: string; firsts: number }[];

    if (rows.length === 0) {
      await this.sendReply(ctx.roomId, ctx.eventId, "No first-poster data yet.");
      return;
    }

    const lines = rows.map((r, i) => `${i + 1}. ${r.user_id} — ${r.firsts} first${r.firsts !== 1 ? "s" : ""}`);
    await this.sendMessage(ctx.roomId, `Early Bird Leaderboard:\n${lines.join("\n")}`);
  }
}

File: internal/plugin/streaks.go
package plugin

import (
	"database/sql"
	"fmt"
	"log/slog"
	"strings"
	"time"

	"gogobee/internal/db"

	"maunium.net/go/mautrix"
	"maunium.net/go/mautrix/id"
)

// StreaksPlugin tracks daily activity streaks and first-poster records.
type StreaksPlugin struct {
	Base
}

// NewStreaksPlugin creates a new streaks plugin.
func NewStreaksPlugin(client *mautrix.Client) *StreaksPlugin {
	return &StreaksPlugin{
		Base: NewBase(client),
	}
}

func (p *StreaksPlugin) Name() string { return "streaks" }

func (p *StreaksPlugin) Commands() []CommandDef {
	return []CommandDef{
		{Name: "streak", Description: "Show your current and longest activity streak", Usage: "!streak", Category: "Leveling & Stats"},
		{Name: "firstboard", Description: "Show top first-posters of the day", Usage: "!firstboard", Category: "Leveling & Stats"},
	}
}

func (p *StreaksPlugin) Init() error { return nil }

func (p *StreaksPlugin) OnReaction(_ ReactionContext) error { return nil }

func (p *StreaksPlugin) OnMessage(ctx MessageContext) error {
	if p.IsCommand(ctx.Body, "streak") {
		return p.handleStreak(ctx)
	}
	if p.IsCommand(ctx.Body, "firstboard") {
		return p.handleFirstboard(ctx)
	}

	// Skip tracking for bot commands
	if ctx.IsCommand {
		return nil
	}

	// Passive: track daily activity and first poster
	p.trackDailyActivity(ctx)
	p.trackFirstPoster(ctx)

	return nil
}

func (p *StreaksPlugin) trackDailyActivity(ctx MessageContext) {
	d := db.Get()
	today := time.Now().UTC().Format("2006-01-02")

	_, err := d.Exec(
		`INSERT INTO daily_activity (user_id, date, message_count) VALUES (?, ?, 1)
		 ON CONFLICT(user_id, date) DO UPDATE SET message_count = message_count + 1`,
		string(ctx.Sender), today,
	)
	if err != nil {
		slog.Error("streaks: track daily activity", "err", err)
	}
}

func (p *StreaksPlugin) trackFirstPoster(ctx MessageContext) {
	d := db.Get()
	today := time.Now().UTC().Format("2006-01-02")
	now := time.Now().UTC().Unix()

	// INSERT OR IGNORE — only the first poster for this room+date wins
	_, err := d.Exec(
		`INSERT OR IGNORE INTO daily_first (room_id, date, user_id, timestamp) VALUES (?, ?, ?, ?)`,
		string(ctx.RoomID), today, string(ctx.Sender), now,
	)
	if err != nil {
		slog.Error("streaks: track first poster", "err", err)
	}
}

func (p *StreaksPlugin) handleStreak(ctx MessageContext) error {
	target := ctx.Sender
	args := p.GetArgs(ctx.Body, "streak")
	if args != "" {
		if resolved, ok := p.ResolveUser(args, ctx.RoomID); ok {
			target = resolved
		}
	}

	d := db.Get()

	// Get all active dates for this user, sorted descending
	rows, err := d.Query(
		`SELECT date FROM daily_activity WHERE user_id = ? ORDER BY date DESC`,
		string(target),
	)
	if err != nil {
		slog.Error("streaks: query", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to look up streak.")
	}
	defer rows.Close()

	var dates []string
	for rows.Next() {
		var date string
		if err := rows.Scan(&date); err != nil {
			continue
		}
		dates = append(dates, date)
	}

	if len(dates) == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("%s has no activity recorded yet.", string(target)))
	}

	currentStreak, longestStreak := calculateStreaks(dates)

	msg := fmt.Sprintf(
		"🔥 %s — Streak Report\nCurrent streak: %s days\nLongest streak: %s days\nTotal active days: %s",
		string(target),
		formatNumber(currentStreak),
		formatNumber(longestStreak),
		formatNumber(len(dates)),
	)
	return p.SendReply(ctx.RoomID, ctx.EventID, msg)
}

// calculateStreaks computes current and longest streaks from descending-sorted date strings.
func calculateStreaks(dates []string) (current int, longest int) {
	if len(dates) == 0 {
		return 0, 0
	}

	today := time.Now().UTC().Format("2006-01-02")

	// Parse first date — current streak only counts if it includes today or yesterday
	firstDate, err := time.Parse("2006-01-02", dates[0])
	if err != nil {
		return 0, 0
	}

	todayDate, _ := time.Parse("2006-01-02", today)
	daysSinceFirst := int(todayDate.Sub(firstDate).Hours() / 24)
	if daysSinceFirst > 1 {
		// Streak is broken — current streak is 0
		// Still calculate longest
		current = 0
	} else {
		// Count current streak
		current = 1
		for i := 1; i < len(dates); i++ {
			prev, err1 := time.Parse("2006-01-02", dates[i-1])
			curr, err2 := time.Parse("2006-01-02", dates[i])
			if err1 != nil || err2 != nil {
				break
			}
			diff := int(prev.Sub(curr).Hours() / 24)
			if diff == 1 {
				current++
			} else {
				break
			}
		}
	}

	// Calculate longest streak (walk all dates)
	streak := 1
	longest = 1
	for i := 1; i < len(dates); i++ {
		prev, err1 := time.Parse("2006-01-02", dates[i-1])
		curr, err2 := time.Parse("2006-01-02", dates[i])
		if err1 != nil || err2 != nil {
			streak = 1
			continue
		}
		diff := int(prev.Sub(curr).Hours() / 24)
		if diff == 1 {
			streak++
		} else {
			streak = 1
		}
		if streak > longest {
			longest = streak
		}
	}

	if current > longest {
		longest = current
	}

	return current, longest
}

func (p *StreaksPlugin) handleFirstboard(ctx MessageContext) error {
	members := p.RoomMembers(ctx.RoomID)

	d := db.Get()

	rows, err := d.Query(
		`SELECT user_id, COUNT(*) as wins
		 FROM daily_first
		 GROUP BY user_id
		 ORDER BY wins DESC`,
	)
	if err != nil {
		slog.Error("streaks: firstboard query", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to load first-poster board.")
	}
	defer rows.Close()

	var sb strings.Builder
	sb.WriteString("🌅 First Poster Board — Top 10\n\n")

	medals := []string{"🥇", "🥈", "🥉"}
	i := 0
	for rows.Next() && i < 10 {
		var userID string
		var wins int
		if err := rows.Scan(&userID, &wins); err != nil {
			continue
		}
		if members != nil && !members[id.UserID(userID)] {
			continue
		}
		prefix := fmt.Sprintf("#%d", i+1)
		if i < len(medals) {
			prefix = medals[i]
		}
		sb.WriteString(fmt.Sprintf("%s %s — %s first posts\n", prefix, userID, formatNumber(wins)))
		i++
	}

	if i == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, "No first-poster data yet.")
	}

	// Also show today's first poster
	today := time.Now().UTC().Format("2006-01-02")
	var todayFirst sql.NullString
	_ = d.QueryRow(
		`SELECT user_id FROM daily_first WHERE room_id = ? AND date = ?`,
		string(ctx.RoomID), today,
	).Scan(&todayFirst)

	if todayFirst.Valid {
		sb.WriteString(fmt.Sprintf("\nToday's first poster in this room: %s", todayFirst.String))
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, sb.String())
}
